xarray/conventions.py

```diff
@@ -155,10 +158,24 @@ def decode_cf_datetime(num_dates, units, calendar=None):
         pd.to_timedelta(flat_num_dates.min(), delta) + ref_date
         pd.to_timedelta(flat_num_dates.max(), delta) + ref_date
 
-        dates = (pd.to_timedelta(flat_num_dates, delta) + ref_date).values
+        # Cast input dates to integers of nanoseconds because `pd.to_datetime`
+        # works much faster when dealing with integers
+        ns_per_time_delta = {'us': 1e3,
+                             'ms': 1e6,
+                             's': 1e9,
+                             'm': 1e9 * 60,
+                             'h': 1e9 * 60 * 60,
+                             'D': 1e9 * 60 * 60 * 24}
```

Comment:
I left it there because the performance impact should be negligible, but it could of course be placed somewhere outside to create it only once.